File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.js
function strStr(haystack, needle) {
    if (needle === "") return 0;


    const lps = Array(needle.length).fill(0);
    let len = 0, i = 1;
    while (i < needle.length) {
        if (needle[i] === needle[len]) {
            lps[i++] = ++len;
        } else if (len > 0) {
            len = lps[len - 1];
        } else {
            i++;
        }
    }
    let h = 0, n = 0;
    while (h < haystack.length) {
        if (haystack[h] === needle[n]) {
            h++; n++;
            if (n === needle.length) return h - n;
        } else if (n > 0) {
            n = lps[n - 1];
        } else {
            h++;
        }
    }
    return -1;
}

File: 0028-find-the-index-of-the-first-occurrence-in-a-string/0028-find-the-index-of-the-first-occurrence-in-a-string.go
func buildFailure(p string) []int {
	fail := make([]int, len(p))
	j := 0
	for i := 1; i < len(p); i++ {
		for j > 0 && p[i] != p[j] {
			j = fail[j-1]
		}
		if p[i] == p[j] {
			j++
		}
		fail[i] = j
	}
	return fail
}

func strStr(haystack, needle string) int {
	if len(needle) == 0 {
		return 0
	}
	fail := buildFailure(needle)
	j := 0
	for i := 0; i < len(haystack); i++ {
		for j > 0 && haystack[i] != needle[j] {
			j = fail[j-1]
		}
		if haystack[i] == needle[j] {
			j++
		}
		if j == len(needle) {
			return i - len(needle) + 1
		}
	}
	return -1
}